tcg/tcg-op.c

```diff
@@ -2829,9 +2829,25 @@ static inline void plugin_gen_mem_callbacks(TCGv_cap_checked_ptr vaddr, uint16_t
 #endif
 }
 
+#ifdef TARGET_AARCH64
+#define CHERI_INVALIDATE_RESPECT_IDX
+#endif
+
 #ifdef TARGET_CHERI
-static inline void gen_cheri_invalidate_tags(TCGv_cap_checked_ptr out_addr, TCGv_i32 memop) {
+static inline void gen_cheri_invalidate_tags(TCGv_cap_checked_ptr out_addr,
+                                             TCGv_i32 memop, TCGArg idx)
+{
+#ifndef CHERI_INVALIDATE_RESPECT_IDX
+    /*
+     * If idx always matches whatever would be used by calling cpu_mmu_index(),
+     * there is no point of including the extra argument on this hot path.
+     */
     gen_helper_cheri_invalidate_tags(cpu_env, out_addr, memop);
```

Comment:
Should it be an error to ever use this on AArch64? If so, perhaps it shouldn't be available there?

Reply:
To use what sorry? gen_helper_cheri_invalidate_tags? No, it is fair any case where tcg_idx was what cpu_mmu_index would return. Just no (obvious) way of knowing in target-general way at translation time that this is the case, unless it is always true (which I think it is for RISCV and MIPS). The only time I know this actual happens is the family of arm instructions that make accesses as if they were made at EL0.

Reply:
I'd have to check, but in some cases it might make be sense to call the helper without the explicit argument for Morello since it avoids the extra tcg temp. CC: @LawrenceEsswood

Reply:
It would, but then then we would either need to pass a flag that indicates an access is to a default mmu_idx (ugly, messy refactor), or somehow have a default mmu_idx stored in the translation context to compare to. Arm already has a default mmu_idx in it's DisasContext, but DisasContextBase (which is what we have access to here) does not. I don't think we want to either. We could, of course, add a 'is_default_mmu_idx' to a number of target specific headers (which can get DisasContext with a bit of container of magic), and have it return false for targets that do not have such a notion, or true if its always true. Seems like a mess to avoid an extra constant load in tcg.

Reply:
Did you measure how much it slows down non-AArch64? Maybe we should just always pass it?

Reply:
Fixed in latest version